### PR TITLE
Use hive column index to parse text file fields

### LINKED ISSUE #2738: Column matching problem of hive external table with csv format
### Steps to reproduce the behavior (Required)
For hive external table with CSV format, the column content display is based on the column order in the text file, not the corresponding content based on the field name of the hive table

### Expected behavior (Required)
The content of the hive external table with CSV format field is the content of the field in the corresponding hive table.

### Real behavior (Required)
For hive external table with CSV format, the column content display is based on the column order in the text file.

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+-------------------+
| current_version() |
+-------------------+
| UNKNOWN e8f11970  |
+-------------------+
1 row in set (0.01 sec)
```

### LINKED ISSUE #2971: BE crash when query from hive external table with csv format
### Steps to reproduce the behavior (Required)
1. create hive partition table and insert some data:
```
create table hive_hdfs_par_csv(
 col_tinyint tinyint comment "column tinyint"
,col_int int comment "column int"
,col_integer integer comment "column integer"
,col_bigint bigint comment "column bigint"
,col_float float comment "column float"
,col_double double comment "column double"
,col_double_precision double precision comment "column double precision"
,col_decimal decimal comment "column decimal"
,col_timestamp timestamp comment "column timestamp"
,col_string string comment "column string"
,col_varchar varchar(100) comment "column varchar"
,col_binary binary comment "column binary"
,col_char char(100) comment "column char"
,col_boolean boolean comment "column boolean"
,col_array array<int> comment "column array"
)
partitioned by(col_date date,col_smallint smallint)
clustered by(col_int) sorted by(col_int) into 32 buckets
row format delimited fields terminated by ','
stored as textfile;
insert into hive_hdfs_par_csv partition(col_date = '2021-01-01',col_smallint=1) select 1,1,1,1,1.001,10.1,110.1,1110.1,'2021-10-30 12:10:23','hello world','hi hello','1110001010101011001001','nihao',true,array(1,10,100);
insert into hive_hdfs_par_csv partition(col_date = '2021-02-01',col_smallint=2) select 2,2,2,2,2.001,20.1,210.1,2110.2,'2021-02-01 12:10:23','hello world','hi hello','1110001010101011001001','nihao',true,array(2,20,200);
insert into hive_hdfs_par_csv partition(col_date = '2021-03-01',col_smallint=3) select 3,3,3,3,3.001,30.1,310.1,3110.1,'2021-03-01 12:10:23','hello world','hi abc','1110001010101011001001','nihao',true,array(3,30,300);
insert into hive_hdfs_par_csv partition(col_date = '2021-04-01',col_smallint=4) select 4,4,4,4,4.001,40.1,410.1,4110.2,'2021-04-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true,array(4,40,400);
insert into hive_hdfs_par_csv partition(col_date = '2021-05-01',col_smallint=5) select 5,5,5,5,5.001,50.1,510.1,5110.1,'2021-05-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true,array(5,50,500);
insert into hive_hdfs_par_csv partition(col_date = '2021-06-01',col_smallint=6) select 6,6,6,6,6.001,60.1,610.1,6110.2,'2021-06-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true,array(6,60,600);
insert into hive_hdfs_par_csv partition(col_date = '2021-07-01',col_smallint=7) select 7,7,7,7,7.001,70.1,710.1,7110.1,'2021-07-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true,array(7,70,700);
insert into hive_hdfs_par_csv partition(col_date = '2021-08-01',col_smallint=8) select 8,8,8,8,8.001,80.1,810.1,8110.2,'2021-08-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true,array(8,80,800);
insert into hive_hdfs_par_csv partition(col_date = '2021-09-01',col_smallint=9) select 9,9,9,9,9.001,90.1,910.1,9110.1,'2021-09-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true,array(9,90,900);
insert into hive_hdfs_par_csv partition(col_date = '2021-10-01',col_smallint=10) select 10,10,10,10,10.001,100.1,1010.1,10110.2,'2021-10-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true,array(10,100,1000);
insert into hive_hdfs_par_csv partition(col_date = '2021-11-01',col_smallint=11) select 11,11,11,11,11.001,110.1,1110.1,11110.2,'2021-11-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true,array(11,110,1100);
insert into hive_hdfs_par_csv partition(col_date = '2021-12-01',col_smallint=12) select 12,12,12,12,12.001,120.1,1210.1,12110.2,'2021-12-01 12:10:23','hello world','hi','1110001010101011001001','nihao',true,array(12,120,1200);
insert into hive_hdfs_par_csv partition(col_date = '2022-01-01',col_smallint=13) select 13,13,13,13,13.001,130.1,1310.1,null,'2021-01-02 12:10:23',null,'hi',null,null,null,array(13,130,1300);
```

2. create hive external table:
```
create external table ex_hive_tbl3 (         
 col_char char(100) comment "column char"         
 ,col_date date comment "column date"         
 ,col_tinyint tinyint comment "column tinyint"         
 ,col_string varchar(1000) comment "column string"         
 ,col_smallint smallint comment "column smallint"         
 ,col_int int comment "column int"         
 ,col_integer integer comment "column integer"         
 ,col_bigint bigint comment "column bigint"         
 ,col_float float comment "column float"         
 ,col_double double comment "column double"         
 ,col_double_precision double comment "column double precision"         
 ,col_decimal decimal comment "column decimal"         
 ,col_timestamp datetime comment "column timestamp"         
 ,col_varchar varchar(100) comment "column varchar"         
 ,col_binary varchar(1000) comment "column binary"         
 ,col_boolean boolean comment "column boolean")         
 ENGINE=hive         
 properties (             
  "resource" = "hive2",             
  "table" = "hive_hdfs_par_csv",             
  "database" = "hive_table_test");
```

3. query on hive external table:
```
mysql> select * from ex_hive_tbl3 order by col_date limit 3;
+----------+------------+-------------+------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------+-------------+------------+-------------+
| col_char | col_date   | col_tinyint | col_string | col_smallint | col_int | col_integer | col_bigint | col_float | col_double | col_double_precision | col_decimal | col_timestamp | col_varchar | col_binary | col_boolean |
+----------+------------+-------------+------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------+-------------+------------+-------------+
| 1        | 2021-01-01 |           1 | 1          |            1 |      10 |         110 |       1110 |      NULL |       NULL |                 NULL |        NULL | NULL          | true        | 110100   |        NULL |
| 2        | 2021-02-01 |           2 | 2          |            2 |      20 |         210 |       2110 |      NULL |       NULL |                 NULL |        NULL | NULL          | true        | 220200   |        NULL |
| 3        | 2021-03-01 |           3 | 3          |            3 |      30 |         310 |       3110 |      NULL |       NULL |                 NULL |        NULL | NULL          | true        | 330300   |        NULL |
+----------+------------+-------------+------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------+-------------+------------+-------------+
3 rows in set (2.22 sec)

mysql> select * from ex_hive_tbl3 order by col_date limit 3;
ERROR 1064 (HY000): rpc failed, host: 172.26.92.141
```

### Expected behavior (Required)
3 rows returned

### Real behavior (Required)
The first query is ok, the second time the same query BE crash:
```
*** Aborted at 1642664770 (unix time) try "date -d @1642664770" if you are using GNU date ***
PC: @          0x2d89c3e starrocks::vectorized::csv::BooleanConverter::read_string()
*** SIGSEGV (@0x0) received by PID 22842 (TID 0x7fd4595d5700) from PID 0; stack trace: ***
    @          0x3480fe2 google::(anonymous namespace)::FailureSignalHandler()
    @     0x7fd47147d655 os::Linux::chained_handler()
    @     0x7fd471482481 JVM_handle_linux_signal
    @     0x7fd471475678 signalHandler()
    @     0x7fd4709aa630 (unknown)
    @          0x2d89c3e starrocks::vectorized::csv::BooleanConverter::read_string()
    @          0x2d7268a starrocks::vectorized::csv::NullableConverter::read_string()
    @          0x24f4960 starrocks::vectorized::HdfsTextScanner::parse_csv()
    @          0x24f4d96 starrocks::vectorized::HdfsTextScanner::do_get_next()
    @          0x24e8432 starrocks::vectorized::HdfsScanner::get_next()
    @          0x2467258 starrocks::vectorized::HdfsScanNode::_scanner_thread()
    @          0x1c26ca0 starrocks::PriorityThreadPool::work_thread()
    @          0x3428dd7 thread_proxy
    @     0x7fd4709a2ea5 start_thread
    @     0x7fd46fda78dd __clone
    @                0x0 (unknown)

```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+------------------------+
| current_version()      |
+------------------------+
| QA_TEST_MASTER 6ffab37 |
+------------------------+
1 row in set (0.01 sec)
```

### PR DESCRIPTION
**Pull request by caneGuy**
## What type of PR is this：
- [x] bug

## Which issues of this PR fixes ：

Fixes #2971 
This pr will also close #2738 

## Problem Summary(Required) ：

we get field index by slot id which will cause core dump and error result if we do not align external table column names with hive original column names.
After this pr
```
MySQL [hive_test]> select * from ex_hive_tbl3 order by col_date;
+------------------------------------------------------------------------------------------------------+------------+-------------+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+-------------+----------------------------------+-------------+
| col_char                                                                                             | col_date   | col_tinyint | col_string  | col_smallint | col_int | col_integer | col_bigint | col_float | col_double | col_double_precision | col_decimal | col_timestamp       | col_varchar | col_binary                       | col_boolean |
+------------------------------------------------------------------------------------------------------+------------+-------------+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+-------------+----------------------------------+-------------+
| nihao                                                                                                | 2021-01-01 |           1 | hello world |            1 |       1 |           1 |          1 |     1.001 |       10.1 |                110.1 |        1110 | 2021-10-30 12:10:23 | hi hello    | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-02-01 |           2 | hello world |            2 |       2 |           2 |          2 |     2.001 |       20.1 |                210.1 |        2110 | 2021-02-01 12:10:23 | hi hello    | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-03-01 |           3 | hello world |            3 |       3 |           3 |          3 |     3.001 |       30.1 |                310.1 |        3110 | 2021-03-01 12:10:23 | hi abc      | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-04-01 |           4 | hello world |            4 |       4 |           4 |          4 |     4.001 |       40.1 |                410.1 |        4110 | 2021-04-01 12:10:23 | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-05-01 |           5 | hello world |            5 |       5 |           5 |          5 |     5.001 |       50.1 |                510.1 |        5110 | 2021-05-01 12:10:23 | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-06-01 |           6 | hello world |            6 |       6 |           6 |          6 |     6.001 |       60.1 |                610.1 |        6110 | 2021-06-01 12:10:23 | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-07-01 |           7 | hello world |            7 |       7 |           7 |          7 |     7.001 |       70.1 |                710.1 |        7110 | 2021-07-01 12:10:23 | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-08-01 |           8 | hello world |            8 |       8 |           8 |          8 |     8.001 |       80.1 |                810.1 |        8110 | 2021-08-01 12:10:23 | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-09-01 |           9 | hello world |            9 |       9 |           9 |          9 |     9.001 |       90.1 |                910.1 |        9110 | 2021-09-01 12:10:23 | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-10-01 |          10 | hello world |           10 |      10 |          10 |         10 |    10.001 |      100.1 |               1010.1 |       10110 | 2021-10-01 12:10:23 | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-11-01 |          11 | hello world |           11 |      11 |          11 |         11 |    11.001 |      110.1 |               1110.1 |       11110 | 2021-11-01 12:10:23 | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-12-01 |          12 | hello world |           12 |      12 |          12 |         12 |    12.001 |      120.1 |               1210.1 |       12110 | 2021-12-01 12:10:23 | hi          | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| NULL                                                                                                 | 2022-01-01 |          13 | NULL        |           13 |      13 |          13 |         13 |    13.001 |      130.1 |               1310.1 |        NULL | 2021-01-02 12:10:23 | hi          | NULL                             |        NULL |
+------------------------------------------------------------------------------------------------------+------------+-------------+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+-------------+----------------------------------+-------------+
13 rows in set (0.04 sec)

MySQL [hive_test]> select * from ex_hive_tbl3 order by col_date limit 3;
+------------------------------------------------------------------------------------------------------+------------+-------------+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+-------------+----------------------------------+-------------+
| col_char                                                                                             | col_date   | col_tinyint | col_string  | col_smallint | col_int | col_integer | col_bigint | col_float | col_double | col_double_precision | col_decimal | col_timestamp       | col_varchar | col_binary                       | col_boolean |
+------------------------------------------------------------------------------------------------------+------------+-------------+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+-------------+----------------------------------+-------------+
| nihao                                                                                                | 2021-01-01 |           1 | hello world |            1 |       1 |           1 |          1 |     1.001 |       10.1 |                110.1 |        1110 | 2021-10-30 12:10:23 | hi hello    | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-02-01 |           2 | hello world |            2 |       2 |           2 |          2 |     2.001 |       20.1 |                210.1 |        2110 | 2021-02-01 12:10:23 | hi hello    | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
| nihao                                                                                                | 2021-03-01 |           3 | hello world |            3 |       3 |           3 |          3 |     3.001 |       30.1 |                310.1 |        3110 | 2021-03-01 12:10:23 | hi abc      | MTExMDAwMTAxMDEwMTAxMTAwMTAwMQ== |           1 |
+------------------------------------------------------------------------------------------------------+------------+-------------+-------------+--------------+---------+-------------+------------+-----------+------------+----------------------+-------------+---------------------+-------------+----------------------------------+-------------+
3 rows in set (0.04 sec)
```

### CODE REVIEW

**Review by DorianZheng — DISMISSED** [review]
LGTM

**Review by dirtysalt — DISMISSED** [review]

File: be/src/exec/vectorized/hdfs_scanner_text.cpp [hunk 1]
```diff
@@ -188,11 +188,18 @@ Status HdfsTextScanner::parse_csv(int chunk_size, ChunkPtr* chunk) {
 
         bool has_error = false;
         int num_materialize_columns = _scanner_params.materialize_slots.size();
+        if (_scanner_params.hive_column_names->size() > fields.size()) {
```
Comment (wyb) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:191]: The size of `hive_column_names` may be larger than `fields` when new columns are added.
The default value should be filled when querying the extra columns that do not exist in the text file.

**Comment by caneGuy** [comment]
PTAL thanks @imay @dirtysalt @DorianZheng @wyb

File: be/src/exec/vectorized/hdfs_scanner_text.cpp [hunk 2]
```diff
@@ -188,14 +188,32 @@ Status HdfsTextScanner::parse_csv(int chunk_size, ChunkPtr* chunk) {
...
         for (int j = 0; j < num_materialize_columns; j++) {
             int index = _scanner_params.materialize_index_in_chunk[j];
-            const Slice& field = fields[_scanner_params.materialize_slots[j]->id() - 1];
-            options.type_desc = &(_scanner_params.materialize_slots[j]->type());
-            if (!_converters[j]->read_string(_column_raw_ptrs[index], field, options)) {
-                chunk->get()->set_num_rows(num_rows);
-                has_error = true;
-                break;
+            int column_field_index = _columns_index[_scanner_params.materialize_slots[j]->col_name()];
+            Column* column = _column_raw_ptrs[index];
```
Comment (DorianZheng) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:199]: Is size of `_columns_index` always equal to size of `_column_raw_ptrs`? should we check if `index` exceed upper bound of it?

Comment (DorianZheng) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:199]: Is size of `_columns_index` always equal to size of `_column_raw_ptrs`? should we check if `index` exceed upper bound of it?

Comment (DorianZheng) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:199]: Is size of `_columns_index` always equal to size of `_column_raw_ptrs`? should we check if `index` exceed upper bound of it?

File: be/src/exec/vectorized/hdfs_scanner_text.cpp [hunk 3]
```diff
@@ -191,7 +191,7 @@ Status HdfsTextScanner::parse_csv(int chunk_size, ChunkPtr* chunk) {
         int field_size = fields.size();
         if (_scanner_params.hive_column_names->size() != field_size) {
             LOG(WARNING) << strings::Substitute("Size mismatch between hive column $0 names and fields $1!",
```
Comment (wyb) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:193]: There will be a lot of logs when parsing old csv files.
Better to remove or use VLOG.

Comment (wyb) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:193]: There will be a lot of logs when parsing old csv files.
Better to remove or use VLOG.

Comment (wyb) [inline be/src/exec/vectorized/hdfs_scanner_text.cpp:193]: There will be a lot of logs when parsing old csv files.
Better to remove or use VLOG.

**Review by wyb — APPROVED** [review]

**Review by DorianZheng — APPROVED** [review]
LGTM

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**